src/model.py

```diff
@@ -214,71 +204,92 @@ def init_engine(self):
-            self._get_bool_config_param("REPORT_CUSTOM_METRICS")
-            and not aync_engine_args.disable_log_stats
-        ):
-            try:
-                labels = {
-                    "model": self.args["model_name"],
-                    "version": self.args["model_version"],
-                }
-                # Add vLLM custom metrics
-                engine_config = self.llm_engine.engine.model_config
-                self.vllm_metrics = VllmStatLogger(
-                    labels, engine_config.max_model_len, self.logger
-                )
-                self.llm_engine.add_logger("triton", self.vllm_metrics)
-            except pb_utils.TritonModelException as e:
-                if "metrics not supported" in str(e):
-                    # Metrics are disabled at the server
-                    self.logger.log_info("[vllm] Metrics not supported")
-                else:
-                    raise e
+        # Run the engine in a separate thread running the AsyncIO event loop.
+        self._llm_engine = None
+        self._llm_engine_start_cv = threading.Condition()
+        self._llm_engine_shutdown_event = asyncio.Event()
+        self._event_thread = threading.Thread(
+            target=asyncio.run, args=(self._run_llm_engine(),)
+        )
+        self._event_thread.start()
+        with self._llm_engine_start_cv:
+            while self._llm_engine is None:
+                self._llm_engine_start_cv.wait()
+
+        # The 'threading.Thread()' will not raise the exception here should the engine
+        # failed to start, so the exception is passed back via the engine variable.
+        if isinstance(self._llm_engine, Exception):
+            e = self._llm_engine
+            self.logger.log_error(f"[vllm] Failed to start engine: {e}")
+            if self._event_thread is not None:
+                self._event_thread.join()
+                self._event_thread = None
+            raise e
 
-    def _get_bool_config_param(self, param_name: str) -> bool:
-        return (param_name in self.model_config["parameters"]) and (
-            self.model_config["parameters"][param_name]["string_value"].lower()
-            == "true"
+    async def _run_llm_engine(self):
+        # Counter to keep track of ongoing request counts.
+        self._ongoing_request_count = 0
+
+        # Check if metrics are enabled. The ZMQ process cannot be used when metrics are
+        # enabled.
+        self._enable_metrics = (
+            self._get_bool_config_param("REPORT_CUSTOM_METRICS")
+            and not self._aync_engine_args.disable_log_stats
         )
 
-    def setup_lora(self):
-        self.enable_lora = False
+        try:
+            # Start the vLLM engine. The engine lives for the scope of this with
+            # statement.
+            async with build_async_engine_client_from_engine_args(
+                engine_args=self._aync_engine_args,
+                disable_frontend_multiprocessing=self._enable_metrics,
```

Comment:
yes, because the engine interface where the metrics was relying on is no longer exposed over ZMQ multi-process, so the ZMQ has to be disabled when metrics are enabled.